### DISCUSSION TITLE
Possible PDF bug: ImageMagick 6.9.12

### DISCUSSION THREAD

**Discussion by skycamdk** [ImageMagick/ImageMagick#3411] [Development]
Using PHP:
imagick module version | 3.4.4
imagick classes | Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version | ImageMagick 6.9.12-1 Q16 x86_64 2021-02-21 https://imagemagick.org
Imagick using ImageMagick library version | ImageMagick 6.9.12-3 Q16 x86_64 2021-03-06 https://imagemagick.org

When reading certain PDF documents I get an error: Uncaught ImagickException: LengthAndFilesizeDoNotMatch `/tmp/magick-_DQFuzZ_vWzRtwQvjCZNpqAvZld05JSx1' @ error/bmp.c/ReadBMPImage/854 in

However the temp file is not written. ImageMagick writes loads of temp files however the file mentioned in the error message is not there.

When using command line "conversion" tool I get no errors. The versions are not the same so that could be the reason why they don't act the same way:
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-01-06 https://imagemagick.org

My php script:

```
#!/usr/bin/php
<?php
$source_path = "compressed.pdf";
$imagick = new Imagick();
$imagick->readImage($source_path);
?>
```

[compressed.pdf](https://github.com/ImageMagick/ImageMagick/files/6154692/compressed.pdf)

**Comment by snibgo** [comment]
> LengthAndFilesizeDoNotMatch `/tmp/magick-_DQFuzZ_vWzRtwQvjCZNpqAvZld05JSx1' @ error/bmp.c/ReadBMPImage/854

I don't understand how your code, reading a PDF file, can give that message about reading a BMP file.

What versions of Ghostscript do you have?

**Reply by skycamdk** [reply]
I haven't been through the IM code. However I see that just by reading the PDF file IM will make images out of each page in the PDF. I have 30 pages in the test PDF and IM creates 29 temp image files in the tmp folder. All pages except the first page (afaik). It then seams like IM tries to read back the temp image files and that fails as one of the files is not there. The error message "LengthAndFilesizeDoNotMatch" might be a bit misleading as I believe that the real error is that the file is not there.

GD headers Version | 2.3.1
GD library Version | 2.3.2

**Comment by snibgo** [comment]
> GD headers Version | 2.3.1
GD library Version | 2.3.2

I don't know what that means. The version of Ghostscript on my laptop is v9.19. If your versions are older, I suggest you upgrade Ghostscript. Current versions are around 9.50.

IM tells Ghostscript to rasterize the PDF into temporary raster image files. These temporary files are of format PAM or PNG, when the rasterization is to CMYK or sRGB respectively. In your case, it seems the file formats are BMP.

**Reply by skycamdk** [reply]
Oh. Sorry. I misunderstood. The ghostscript version is:
GPL Ghostscript 9.25 (2018-09-13)
Is there a way to debug what is going on inside IM?

**Comment by snibgo** [comment]
At the command line, insert "-debug all" as the first option. I don't know the equivalent in IMagick. (Note that IMagick is a separate product to ImageMagick, that provides an interface between PHP and ImageMagick. They are not maintained by the same people.)

**Comment by fmw42** [comment]
Using ImageMagick 6.9.12.3 Q16 Mac OSX and Ghostscript 9.50, I can do

`convert compressed.pdf x.png`

and get 30 png images and no error messages

So it is either your version of ImageMagick that has a potential bug or your version of Ghostscript or Imagick or your Imagick coding.  I note that GS 9.25 is old and might be around the time there were some bugs in GS. So I would suggest you upgrade your GS and try again.
